kits/daisy/src/robot/hexapod.cpp

```diff
@@ -0,0 +1,610 @@
+  {
+    // Replan trajectory every timestep for legs in flight
+    if (leg->getMode() == Leg::Mode::Flight)
+      leg->updateStep(t);
+  }
+}
+
+Eigen::VectorXd Hexapod::getLegFeedback(int leg_index)
+{
+  Eigen::VectorXd leg_angles(Leg::getNumJoints());
+  int num_joints = Leg::getNumJoints();
+  int leg_offset = leg_index * num_joints;
+
+  std::lock_guard<std::mutex> guard(fbk_lock_);
+
+  for (int i = 0; i < num_joints; ++i)
+    leg_angles[i] = positions_[leg_offset + i];
+
+  return leg_angles;
+}
+
+void Hexapod::computeFootForces(double t, Eigen::MatrixXd& foot_forces)
+{
+  Eigen::VectorXd factors(6);
+  Eigen::VectorXd blend_factors(6);
+  Eigen::Vector3d grav;
+  grav << 0, 0, 1;
+  // Get the dot product of gravity with each leg, and then subtract a scaled
+  // gravity from the foot stance position.
+  // NOTE: Matt is skeptical about this overall approach; but it worked before so we are keeping
+  // it for now.
+  factors.resize(6); 
+  for (int i = 0; i < 6; ++i)
+  {
+    Eigen::Vector3d stance = legs_[i]->getCmdStanceXYZ();
+    double dot_prod = grav.dot(stance);
+    factors(i) = (grav * dot_prod - stance).norm();
+  }
+  double fact_sum = factors.sum();
+  for (int i = 0; i < 6; ++i)
+    factors(i) = fact_sum / factors(i);
+  for (int i = 0; i < 6; ++i)
+  {
+    // Redistribute weight to just modules in stance
+    if (legs_[i]->getMode() == Leg::Mode::Flight)
+    {
+      double switch_time = 0.1; // XML? time where weight is being shifted to/from a foot
+      double curr_step_time = legs_[i]->getStepTime(t);
+      double step_period = legs_[i]->getStepPeriod();
+      if (curr_step_time < switch_time)
+      {
+        blend_factors(i) = (switch_time - curr_step_time) / switch_time;
+      }
+      else if ((step_period - curr_step_time) < switch_time)
+      {
+        blend_factors(i) = (curr_step_time - (step_period - switch_time)) / switch_time;
+      }
+      else
+      {
+        blend_factors(i) = 0;
+      }
+
+      factors(i) *= blend_factors(i);
+    }
+    else
+    {
+      blend_factors(i) = 1;
+    }
+  }
+  fact_sum = factors.sum();
+  factors /= fact_sum;
+
+  // NOTE: here, we have a blend factor for each foot to allow for future gaits;
+  // in MATLAB, there was just one scalar for this.  We use "max" here to match
+  // the results from MATLAB.
+  for (int i = 0; i < 6; ++i)
+    factors(i) = factors(i) * (1 + .33 * std::sin(M_PI * blend_factors(i)));
+
+//  std::cout << "factors: " << factors << std::endl;
+//  std::cout << "grav: " << grav << std::endl;
+
+  foot_forces.resize(3,6);
+  for (int i = 0; i < 6; ++i)
+    foot_forces.block<3,1>(0,i) = factors(i) * weight_ * grav;
+}
+
+// TODO: make this a class function and remove parameters? Make it better!
+bool copyIntoPositions(Eigen::VectorXd& positions, const hebi::GroupFeedback* fbk, const std::set<int>& real_legs)
+{
+  bool valid_fbk = true;
+  int fbk_idx = 0;
+  int num_leg_modules = 3;
+  for (int i = 0; i < positions.size();)
```

Comment:
I really don't understand what's going on here

Reply:
Added a comment to clarify.  Let's discuss -- we could potentially remove all of the "partial number of legs" support to clarify a decent bit of code.